grace/run.py

```diff
@@ -0,0 +1,263 @@
+        input_data = ImageGraphDataset(
+            image_dir=image_dir,
+            grace_dir=grace_dir,
+            image_filetype=config.filetype,
+            keep_node_unknown_labels=config.keep_node_unknown_labels,
+            keep_edge_unknown_labels=config.keep_edge_unknown_labels,
+            transform=transform_method,
+        )
+
+        # Process the (sub)graph data into torch_geometric dataset:
+        target_list, subgraph_dataset = [], []
+        desc = "Extracting patch features from images... "
+        for _, target in tqdm(input_data, desc=desc, disable=not verbose):
+            file_name = target["metadata"]["image_filename"]
+            LOGGER.info(f"Processing file: {file_name}")
+
+            # Store the valid graph list:
+            target_list.append(target)
+
+            # Chop graph into subgraphs & store:
+            graphs = dataset_from_graph(target["graph"], mode=graph_processing)
+            subgraph_dataset.extend(graphs)
+
+        return target_list, subgraph_dataset
+
+    # Create a transform function with frozen 'in_train_mode' parameter:
+    transform_train_mode = partial(transform, in_train_mode=True)
+    transform_valid_mode = partial(transform, in_train_mode=False)
+
+    # Read the respective datasets:
+    _, train_dataset = prepare_dataset(
+        config.train_image_dir,
+        config.train_grace_dir,
+        transform_train_mode,
+    )
+    valid_target_list, valid_dataset = prepare_dataset(
+        config.valid_image_dir,
+        config.valid_grace_dir,
+        transform_valid_mode,
+    )
+    infer_target_list, _ = prepare_dataset(
+        config.infer_image_dir,
+        config.infer_grace_dir,
+        transform_valid_mode,
+    )
+
+    # Define the GNN classifier model:
+    classifier = GNNClassifier().get_model(
+        config.classifier_type,
+        input_channels=config.feature_dim,
+        hidden_channels=config.hidden_channels,
+        dropout=config.dropout,
+        node_output_classes=config.num_node_classes,
+        edge_output_classes=config.num_edge_classes,
+    )
+
+    # Perform the training:
+    train_model(
+        classifier,
+        train_dataset,
+        valid_dataset,
+        valid_target_list,
+        epochs=config.epochs,
+        batch_size=config.batch_size,
+        learning_rate=config.learning_rate,
+        log_dir=run_dir,
+        metrics=config.metrics_classifier,
+        tensorboard_update_frequency=config.tensorboard_update_frequency,
+        valid_graph_ploter_frequency=config.valid_graph_ploter_frequency,
+    )
+
+    # Save the trained model:
+    model_save_fn = run_dir / "classifier.pt"
+    torch.save(classifier, model_save_fn)
+    write_config_file(config, filetype="json")
+    write_config_file(config, filetype="yaml")
+
+    # Project the TSNE manifold:
+    if config.visualise_tsne_manifold is True:
+        # TODO: Implement
+        pass
```

Comment:
maybe add a NotImplemented warning here?

Reply:
Coming in the next commit 🚀